govcd/defined_interface.go

```diff
@@ -192,6 +192,153 @@ func (di *DefinedInterface) Delete() error {
+// Only allowed if the Interface is not in use.
+func (di *DefinedInterface) AddBehavior(behavior types.Behavior) (*types.Behavior, error) {
+	if di.DefinedInterface.ID == "" {
+		return nil, fmt.Errorf("ID of the receiver Defined Interface is empty")
+	}
+
+	endpoint := types.OpenApiPathVersion1_0_0 + types.OpenApiEndpointRdeInterfaceBehaviors
+	apiVersion, err := di.client.getOpenApiHighestElevatedVersion(endpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	urlRef, err := di.client.OpenApiBuildEndpoint(fmt.Sprintf(endpoint, di.DefinedInterface.ID))
+	if err != nil {
+		return nil, err
+	}
+
+	result := &types.Behavior{}
+	err = di.client.OpenApiPostItem(apiVersion, urlRef, nil, behavior, result, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return result, nil
+}
+
+// GetAllBehaviors retrieves all the Behaviors of the receiver Defined Interface. Query parameters can be supplied to modify pagination.
+func (di *DefinedInterface) GetAllBehaviors(queryParameters url.Values) ([]*types.Behavior, error) {
+	if di.DefinedInterface.ID == "" {
+		return nil, fmt.Errorf("ID of the receiver Defined Interface is empty")
+	}
+	return getAllBehaviors(di.client, di.DefinedInterface.ID, types.OpenApiEndpointRdeInterfaceBehaviors, queryParameters)
+}
+
+// getAllBehaviors gets all the Behaviors from the object referenced by the input Object ID with the given OpenAPI endpoint.
+func getAllBehaviors(client *Client, objectId, openApiEndpoint string, queryParameters url.Values) ([]*types.Behavior, error) {
+	endpoint := types.OpenApiPathVersion1_0_0 + openApiEndpoint
+	apiVersion, err := client.getOpenApiHighestElevatedVersion(endpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	urlRef, err := client.OpenApiBuildEndpoint(fmt.Sprintf(endpoint, objectId))
+	if err != nil {
+		return nil, err
+	}
+
+	typeResponses := []*types.Behavior{{}}
+	err = client.OpenApiGetAllItems(apiVersion, urlRef, queryParameters, &typeResponses, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return typeResponses, nil
+}
+
+// GetBehaviorById retrieves a unique Behavior that belongs to the receiver Defined Interface and is determined by the
+// input ID.
+func (di *DefinedInterface) GetBehaviorById(id string) (*types.Behavior, error) {
+	endpoint := types.OpenApiPathVersion1_0_0 + types.OpenApiEndpointRdeInterfaceBehaviors
+	apiVersion, err := di.client.getOpenApiHighestElevatedVersion(endpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	urlRef, err := di.client.OpenApiBuildEndpoint(fmt.Sprintf(endpoint, di.DefinedInterface.ID), id)
+	if err != nil {
+		return nil, err
+	}
+
+	response := types.Behavior{}
+	err = di.client.OpenApiGetItem(apiVersion, urlRef, nil, &response, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return &response, nil
+}
+
+// GetBehaviorByName retrieves a unique Behavior that belongs to the receiver Defined Interface and is named after
+// the input.
+func (di *DefinedInterface) GetBehaviorByName(name string) (*types.Behavior, error) {
```

Comment:
Similar question to the one for defined entity - can there be duplicate names? Any chance it can filter on API (I suspect not)?

Reply:
Answered in a similar thread: https://github.com/vmware/go-vcloud-director/pull/584#discussion_r1260738753